### ISSUE TITLE
Expression sub-language for modifying state

### ISSUE THREAD

**Issue by jeromesimeon** [accordproject/ergo#143]
When changing the state, one has currently to rebuild the whole state each time. It might be useful to have expressions specifically designed to modify the state. For instance:

```
set status to "BREACHED" in state;
add new PaymentObligation to obligations in state;
remove notice in obligations from state;
return ...
```

**Comment by mttrbrts** [comment]
How about something like ... 
```set state.status to "BREACHED";
add new PaymentObligation to state.obligations;
remove notice from state.obligations;
return,```

**Comment by jeromesimeon** [comment]
I like it! One issue is what does `remove` mean, for which there are several options:
- is remove based on value equality (i.e., notice is an obligation object and your compare the structure)
- is remove based on a kind of identity (we never use identifiers anywhere at the moment)
- is remove based on another kind of criteria

We could imagine something like:
```
remove x from state.obligation where x.kind = "notice";
remove x from state.obligation where x.id = "5555-3562-1412";
return ...
```

**Comment by jeromesimeon** [comment]
Another question, is whether those new operations should replace the current `set state` statement altogether (i.e., remove the current `set state`).

**Comment by dselman** [comment]
For types that have `identified by` in the model (basically everything except concepts and enums) I think it would be natural to use the modelled identifier for equality. For concepts I guess we either have to use value-based equality, or allow the user to specify an `identified by` expression to the call to `remove`.

I do like the `where` expressions for selection as well however.

How about `update state` to contrast with `set state`?

**Comment by jeromesimeon** [comment]
For the records, @kach had suggested the nice and compact:
```
state.status <- BREACHED;
```

And add/remove can technically be handled using array operations:
```
state.obligations <- arrayAdd(state.obligations, [new PaymentObligation{ ... }]);
state.obligations <- arraySubtract(state.obligations, notice);
```

**Comment by mttrbrts** [comment]
I like the brevity of this suggestion, but would rather avoid introducing another operator.

With this in mind what about ... ?
```
set state.status = BREACHED;
```

This aligns the syntax for state assignment with assignment in let expressions.

**Comment by hm2space** [comment]
There is a nice syntax in F# (OCaml Dotnet variant) that does something like this 
set state { state with status = "STARTED" }
It is aligned with the "immutability" mindset.
the expression is like 
{ objReference 'with' fieldAssignmentList }

**Comment by jeromesimeon** [comment]
I like the idea of conveying cloning since this could be used outside of setting state. I don’t know why I didn’t think of this. And of course the ocaml/f# syntax is quite nice for that.

Note that things are a bit more difficult in Ergo since we need those operations to be aware of the “class hierarchy”, not just records as in OCaml.

A branch is tracking this, but I still need to work through the typing ramifications.

**Comment by hm2space** [comment]
Is there a level of meta-programming or reflection on types ?

**Comment by jeromesimeon** [comment]
Not in Ergo itself if I understand your question correctly. Although you can use match expression on type names which are declared as part of the hierarchy.

**Comment by jeromesimeon** [comment]
In JavaScript:

```
{ ...state, "status": "BREACHED" }
```

**Comment by HamidMoutawakkil** [comment]
Nice. You can even save typing by using & instead

**Comment by dselman** [comment]
> In JavaScript, we can use the spread operator for cloning. This might be a good syntax in Ergo as well?
> 
> ```
> set state { ...state, "status": "BREACHED" }
> ```

Doesn't seem that obvious to a less technical user. `update state` ?

**Comment by jeromesimeon** [comment]
@dselman that was also the impression on the Ergo call this morning.

I do find a syntax for cloning appealing since it can be applied to things other than the `state`, but it's maybe something to wait on.

**Comment by HamidMoutawakkil** [comment]
The logic is to be able to clone and override any type not just the state. 
let myObj = {...someBojOfSameType, overridenFieldList }

state being here just one case of it.

**Comment by jeromesimeon** [comment]
> I like the brevity of this suggestion, but would rather avoid introducing another operator.
> 
> With this in mind what about ... ?
> 
> ```
> set state.status = BREACHED;
> ```
> 
> This aligns the syntax for state assignment with assignment in let expressions.

This is the proposal that has been implemented in #667 

Note that this new feature is specific to the state, this does not include support for things like spread operator.

**Comment by jeromesimeon** [comment]
Implemented in the `0.20` branch. Tested in Cicero and Cicero template library.
